### ISSUE TITLE
Importer - Rule Order issue with moved rules

### ISSUE THREAD

**Issue by tpurschke** [CactuseSecurity/firewall-orchestrator#2761]
In  R001 production system the standard rules report does not show the correct order of rules.

Root cause: the calculation of rule order during import process does not work for rule moves!

Need to
- [ ] rework/enhance existing ruleorder unit tests to verify correct rule order on inserts, deletes and moves in  updateRulebaseDiffs
- [x] integrate compute_min_moves in updateRulebaseDiffs
- [x] fix this (in Importer rework) - move must actually create new rule version. 
- [x] update occurences of old rule versions id as fk
- [x] write unit test for compute_min_moves in the context of inserting, deleting and moving rules
- [x] hierarchy needs to be reflected correctly (Levels go deeper than one Level), rulebases can start in the middle of another rulebase
- [x] fix this once at customer by running clear import

Current approach: "universal rule order algorithm" + absolute order number in rule_num_numeric

**Comment by Robin-Smets** [comment]
To solve task two ("fix this (in Importer rework) - move must actually create new rule version"), I would need to ask Ralf to move a rule — possibly one created specifically for this purpose — in order to verify whether the reordering would actually work in a production scenario, right?

**Comment by tpurschke** [comment]
> To solve task two ("fix this (in Importer rework) - move must actually create new rule version"), I would need to ask Ralf to move a rule — possibly one created specifically for this purpose — in order to verify whether the reordering would actually work in a production scenario, right?

yes, but I can also do this easily. Just tell me, when you need it.

**Comment by Robin-Smets** [comment]
I’ve just finished testing. As shown in the screenshots below, the updated order arrays after a change behave as expected. However, the change counter is misleading, as there were clearly modifications in the database.
Additionally, since the logic is about to be updated to reflect the rulebase hierarchy, I suggest we repeat the same test once those changes are in place.

**Data after import before changing the order**

![Image](https://github.com/user-attachments/assets/f27de56f-262f-4b62-b266-78e043511c63)

**Data after second import after changing the order**

![Image](https://github.com/user-attachments/assets/b1ea669a-bac6-40cc-a038-11f1b3c45c3a)

**Terminal output after second import**

![Image](https://github.com/user-attachments/assets/190a7f37-8731-4c63-bf69-12d28b5b7fe3)

**Comment by Robin-Smets** [comment]
I am sorry, but I can't see a new approach here. I would be very happy if you can correct me on this, but here are my observations and thoughts:

What the method does is reducing the amount of updates, but we still need to update (depending on the case) a lot of rules if we move a rule.

I try to explain what I mean with an example. Consider the case we did the first import and now have a second import which has moved rules

Previous Config (first import)

rule 1 - rule_num_numeric 1
rule 2 - rule_num_numeric 2
rule 3 - rule_num_numeric 3
rule 4 - rule_num_numeric 4
rule 5 - rule_num_numeric 5

Current Config (second import)

rule 1 - rule_num_numeric 1
rule 5 - rule_num_numeric 5
rule 2 - rule_num_numeric 2
rule 3 - rule_num_numeric 3
rule 4 - rule_num_numeric 4

The method would tell me, correctly that rule 5 moved to index 1. But to change the order I have to...

a) Change rule 5, 2, 3, 4. That was my first approach, when I just made new rule_num_numerics for every rule in the config. The difference would only be, that we can recognize that we dont have to start with the first rule, but we can start with the second.

b) Have some kind of rule order table, which was the second approach we agreed on.

c) Insert with fractions. So it would be something like getting the current rule_num_numeric on that position and the position before that and calculating a float that is between those two values. It would result in something like this:

rule 1 - rule_num_numeric 1
rule 5 - rule_num_numeric 1.5
rule 2 - rule_num_numeric 2
rule 3 - rule_num_numeric 3
rule 4 - rule_num_numeric 4

This approach would only need to update one line per move. Personally I dont like that approach. If we have a lot of moves we can run into floating point precision issues and it is an approach that seems to me an approach that will attract bugs as electric light. 

Please tell me if I miss something here and which way we should go.

**Comment by tpurschke** [comment]
Hi Robin, c) is exactly what I have in mind (and what we have already implemented before.
It is the reason why the precision of the rule_num_numeric field is set so high.
Maybe we can discuss the bug/light scenario tomorrow ;-)

**Comment by tpurschke** [comment]
current state of importer regarding rule move:
- all rule table entries seem fine
- rule_from, rule_to, rule_svc, ... do not get updated when moving a rule
- this results in empty src/dst/svc in report when showing moved rules
- also the rule did no get placed in the correct position but  one rule further down

#### firewall view:
![Image](https://github.com/user-attachments/assets/e177dc30-0d8d-44e7-97de-218f4a214e95)

#### fwo report view:

![Image](https://github.com/user-attachments/assets/036c8ac7-b303-4e01-b727-bf6f29018326)

#### rule num database situation:
![Image](https://github.com/user-attachments/assets/0b0924eb-a709-47fa-8f00-6f41f0cd924b)

**Comment by Robin-Smets** [comment]
> current state of importer regarding rule move:
> 
>     * all rule table entries (except rule_num_numeric for the new version of the rule) seem fine
> 
>     * rule_from, rule_to, rule_svc, ... do not get updated when moving a rule
> 
>     * this results in empty src/dst/svc in report when showing moved rules
> 
>     * also the rule did no get placed in the correct position but  one rule further down (has the same, not a lower rule_num_numeric value as the rule which it should be placed before)
> 


1. I fixed the los src/dst/svcs. But I am not sure if that suffices. I searched for "rule_id" in the sql script for table creation, to find other places where updates might be necessary, but I am not sure, because that depends strongly on the use case. Maybe you can have a look on the list down below and tell me which tables should be updated, because they refer always to the current version of the rule and which are more in the historic context or in a technical context that demands the fk to stay like it is set initially.
2. I had to make some changes to the api call (fwo_api_oo.py). Then sonarqube demanded a greater refactoring of the methods to reduce complexity. There were no problems during my tests, but it might be good if this passage gets reviewed thoroughly.
3. I could not reproduce the bug about the wrong position of moved rules.

**Comment by tpurschke** [comment]
unfortunately I could reproduce it again. The rule_num_numeric is again the same as that of the next rule:

![Image](https://github.com/user-attachments/assets/5aeb3954-c5b8-446e-8487-21b9af4583f6)
